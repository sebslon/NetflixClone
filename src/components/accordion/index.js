import React, { createContext, useState, useContext } from "react";

import {
  Body,
  Container,
  Header,
  Item,
  Inner,
  Title,
} from "./styles/accordion";

const ToggleContext = createContext();

export default function Accordion({ children, ...otherProps }) {
  return (
    <Container {...otherProps}>
      <Inner>{children}</Inner>
    </Container>
  );
}

Accordion.Title = function AccordionTitle({ children, ...otherProps }) {
  return <Title {...otherProps}>{children}</Title>;
};

Accordion.Item = function AccordionItem({ children, ...otherProps }) {
  const [toggleShow, setToggleShow] = useState(false);

  return (
    <ToggleContext.Provider value={{ toggleShow, setToggleShow }}>
      <Item {...otherProps}>{children}</Item>
    </ToggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...otherProps }) {
  const { toggleShow, setToggleShow } = useContext(ToggleContext);

  return (
    <Header
      onClick={() => setToggleShow((prevState) => !prevState)}
      {...otherProps}
    >
      {children}
      {toggleShow ? (
        <img src="images/icons/close-slim.png" alt="close" />
      ) : (
        <img src="images/icons/add.png" alt="open" />
      )}
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...otherProps }) {
  const { toggleShow } = useContext(ToggleContext);

  return toggleShow ? <Body {...otherProps}>{children}</Body> : null;
};
