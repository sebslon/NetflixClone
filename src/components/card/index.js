import { createContext, useContext, useState } from "react";

import {
  Container,
  Content,
  Entities,
  Feature,
  FeatureClose,
  FeatureTitle,
  FeatureText,
  Group,
  Item,
  Image,
  Maturity,
  Meta,
  SubTitle,
  Title,
  Text,
} from "./styles/card";

export const FeatureContext = createContext();

export default function Card({ children, ...otherProps }) {
  const [showFeature, setShowFeature] = useState(false);
  const [itemFeature, setItemFeature] = useState({});

  return (
    <FeatureContext.Provider
      value={{ showFeature, setShowFeature, itemFeature, setItemFeature }}
    >
      <Container {...otherProps}>{children}</Container>
    </FeatureContext.Provider>
  );
}

Card.Group = function CardGroup({ children, ...otherProps }) {
  return <Group {...otherProps}>{children}</Group>;
};

Card.Title = function CardTitle({ children, ...otherProps }) {
  return <Title {...otherProps}>{children}</Title>;
};

Card.SubTitle = function CardSubTitle({ children, ...otherProps }) {
  return <SubTitle {...otherProps}>{children}</SubTitle>;
};

Card.Text = function CardText({ children, ...otherProps }) {
  return <Text {...otherProps}>{children}</Text>;
};

Card.Entities = function CardEntity({ children, ...otherProps }) {
  return <Entities {...otherProps}>{children}</Entities>;
};

Card.Meta = function CardMeta({ children, ...otherProps }) {
  return <Meta {...otherProps}>{children}</Meta>;
};

Card.Feature = function CardFeature({ children, category, ...otherProps }) {
  const { showFeature, itemFeature, setShowFeature } = useContext(FeatureContext);

  return showFeature ? (
    <Feature {...otherProps}
      src={`/images/${category}/${itemFeature.genre}/${itemFeature.slug}/large.jpg`}
    >
      <Content>
        <FeatureTitle>{itemFeature.title}</FeatureTitle>
        <FeatureText>{itemFeature.description}</FeatureText>
        <FeatureClose onClick={() => setShowFeature(false)}>
          <img src="/images/icons/close.png" alt="close" />
        </FeatureClose>

        <Group margin="30px 0" flexDirection="row" alignItems="center">
          <Maturity rating={itemFeature.maturity}>
            {itemFeature.maturity > 12 ? "PG" : itemFeature.maturity}
          </Maturity>
          <FeatureText fontWeight="bold">
            {itemFeature.genre.charAt(0).toUpperCase() + itemFeature.genre.slice(1)}
          </FeatureText>
        </Group>
        {children}
      </Content>
    </Feature>
  ) : null;
};

Card.Item = function CardItem({ item, children, ...otherProps }) {
  const { setShowFeature, setItemFeature } = useContext(FeatureContext);

  return (
    <Item
      onClick={() => {
        setItemFeature(item);
        setShowFeature(true);
      }}
      {...otherProps}
    >
      {children}
    </Item>
  );
};

Card.Image = function CardImage({ ...otherProps }) {
  return <Image {...otherProps} />;
};
