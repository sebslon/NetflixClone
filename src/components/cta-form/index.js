import { Break, Container, Input, Button, Text } from './styles/cta-form';

export default function CtaForm({ children, ...otherProps }) {
  return (
    <Container {...otherProps}>{children}</Container>
  )
}

CtaForm.Input = function CtaFormInput({ ...otherProps }) {
  return <Input {...otherProps} />
}

CtaForm.Button = function CtaFormButton({ children, ...otherProps }) {
  return (
    <Button {...otherProps}>
      {children} <img src="images/icons/chevron-right.png" alt="Try now" />
    </Button>
  )
};

CtaForm.Text = function CtaFormText({ children, ...otherProps }) {
  return <Text {...otherProps}>{children}</Text>
}

CtaForm.Break = function CtaFormBreak({...otherProps}) {
  return <Break {...otherProps} />;
}