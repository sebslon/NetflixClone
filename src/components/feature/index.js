import { Container, Title, SubTitle } from './styles/feature';

export default function Feature({children, ...otherProps}) {
  return <Container {...otherProps}>{children}</Container>
};

Feature.Title = function FeatureTitle({children, ...otherProps}) {
  return <Title {...otherProps}>{children}</Title>
}

Feature.SubTitle = function FeatureSubTitle({children, ...otherProps}) {
  return <SubTitle {...otherProps}>{children}</SubTitle>
}