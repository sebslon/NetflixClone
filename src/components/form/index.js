import { Base, Container, Error, Title, Text, TextSmall, Link, Input, Submit } from './styles/form';

export default function Form({ children, ...otherProps }) {
  return <Container {...otherProps}>{children}</Container>
}

Form.Base = function FormBase({ children, ...otherProps }) {
  return <Base {...otherProps}>{children}</Base>
}

Form.Error = function FormError({ children, ...otherProps }) {
  return <Error {...otherProps}>{children}</Error>
}

Form.Title = function FormTitle({ children, ...otherProps }) {
  return <Title {...otherProps}>{children}</Title>
}

Form.Text = function FormText({ children, ...otherProps }) {
  return <Text {...otherProps}>{children}</Text>
}

Form.TextSmall = function FormTextSmall({ children, ...otherProps }) {
  return <TextSmall {...otherProps}>{children}</TextSmall>
}

Form.Link = function FormLink({ children, ...otherProps }) {
  return <Link {...otherProps}>{children}</Link>
}

Form.Input = function FormInput({ children, ...otherProps }) {
  return <Input {...otherProps}>{children}</Input>
}

Form.Submit = function FormSubmit({ children, ...otherProps }) {
  return <Submit {...otherProps}>{children}</Submit>
}