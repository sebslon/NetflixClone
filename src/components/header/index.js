import { Link } from 'react-router-dom';

import { Background, Container, ButtonLink, Logo } from './styles/header';

export default function Header({ background = true, children, ...otherProps }) {
  return background ? <Background {...otherProps}>{children}</Background> : children;
}

Header.Frame = function HeaderFrame({ children, ...otherProps }) {
  return <Container {...otherProps}>{children}</Container>;
}

Header.ButtonLink = function HeaderButtonLink({ children, ...otherProps }) {
  return <ButtonLink {...otherProps}>{children}</ButtonLink>;
}

Header.Logo = function HeaderLogo({ to, ...otherProps }) {
  return (
    <Link to={to}>
      <Logo {...otherProps} />
    </Link>
  )
}
