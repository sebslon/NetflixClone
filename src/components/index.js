export { default as Accordion } from './accordion'
export { default as CtaForm } from './cta-form'
export { default as Footer } from './footer';
export { default as Jumbotron } from './jumbotron';
