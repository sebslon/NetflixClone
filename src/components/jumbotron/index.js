import React from 'react';

import { Container, Inner, Item, Pane, Title, SubTitle, Image } from './styles/jumbotron';

export default function Jumbotron({ children, direction = 'row', ...otherProps }) {
  return (
    <Item>
      <Inner direction={direction}>
        {children}
      </Inner>
    </Item>
  )
}

Jumbotron.Container = function JumbotronContainer({ children, ...otherProps }) {
  return <Container {...otherProps}>{children}</Container>
}

Jumbotron.Pane = function JumbotronPane({ children, ...otherProps }) {
  return <Pane {...otherProps}>{children}</Pane>
}

Jumbotron.Title = function JumbotronTitle({ children, ...otherProps }) {
  return <Title {...otherProps}>{children}</Title>
}

Jumbotron.SubTitle = function JumbotronSubTitle({ children, ...otherProps }) {
  return <SubTitle {...otherProps}>{children}</SubTitle>
}

Jumbotron.Image = function JumbotronImage({ ...otherProps }) {
  return <Image {...otherProps} />
}