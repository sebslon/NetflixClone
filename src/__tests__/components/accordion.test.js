import { queryByText, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import faqData from "../../fixtures/faqs.json";
import { Accordion } from "../../components";

describe("<Accordion />", () => {
  it("Renders <Accordion /> with populated data", () => {
    const { container, getByText } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        {faqData.map((item) => (
          <Accordion.Item key={item.id}>
            <Accordion.Header>{item.header}</Accordion.Header>
            <Accordion.Body>{item.body}</Accordion.Body>
          </Accordion.Item>
        ))}
      </Accordion>
    );

    expect(getByText("Frequently Asked Questions")).toBeTruthy();
    expect(getByText("What is Netflix?")).toBeTruthy();
    expect(getByText("Where can I watch?")).toBeTruthy();
    expect(getByText("How do I cancel?")).toBeTruthy();
    expect(getByText("What can I watch on Netflix?")).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Opens and closes the <Accordion /> component when clicked", () => {
    const { container, queryByText } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        {faqData.map((item) => (
          <Accordion.Item key={item.id}>
            <Accordion.Header>{item.header}</Accordion.Header>
            <Accordion.Body>{item.body}</Accordion.Body>
          </Accordion.Item>
        ))}
      </Accordion>
    );

    const faqText =
      "Netflix has an extensive library of feature films, documentaries, TV programmes, anime, award-winning Netflix originals, and more. Watch as much as you want, any time you want.";

    expect(queryByText(faqText)).toBeFalsy();

    userEvent.click(queryByText("What can I watch on Netflix?"));
    expect(queryByText(faqText)).toBeTruthy();

    userEvent.click(queryByText("What can I watch on Netflix?"));
    expect(queryByText(faqText)).toBeFalsy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
