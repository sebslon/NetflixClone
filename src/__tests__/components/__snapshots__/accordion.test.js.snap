// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> Opens and closes the <Accordion /> component when clicked 1`] = `
<div
  class="accordion__Container-l3iym9-0 bAxAgi"
>
  <div
    class="accordion__Inner-l3iym9-1 kXGanl"
  >
    <h1
      class="accordion__Title-l3iym9-5 jFkLSD"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="accordion__Item-l3iym9-2 gwMLOH"
    >
      <div
        class="accordion__Header-l3iym9-4 cIYXTq"
      >
        What is Netflix?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-l3iym9-2 gwMLOH"
    >
      <div
        class="accordion__Header-l3iym9-4 cIYXTq"
      >
        How much does Netflix cost?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-l3iym9-2 gwMLOH"
    >
      <div
        class="accordion__Header-l3iym9-4 cIYXTq"
      >
        Where can I watch?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-l3iym9-2 gwMLOH"
    >
      <div
        class="accordion__Header-l3iym9-4 cIYXTq"
      >
        How do I cancel?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-l3iym9-2 gwMLOH"
    >
      <div
        class="accordion__Header-l3iym9-4 cIYXTq"
      >
        What can I watch on Netflix?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Accordion /> Renders <Accordion /> with populated data 1`] = `
<div
  class="accordion__Container-l3iym9-0 bAxAgi"
>
  <div
    class="accordion__Inner-l3iym9-1 kXGanl"
  >
    <h1
      class="accordion__Title-l3iym9-5 jFkLSD"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="accordion__Item-l3iym9-2 gwMLOH"
    >
      <div
        class="accordion__Header-l3iym9-4 cIYXTq"
      >
        What is Netflix?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-l3iym9-2 gwMLOH"
    >
      <div
        class="accordion__Header-l3iym9-4 cIYXTq"
      >
        How much does Netflix cost?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-l3iym9-2 gwMLOH"
    >
      <div
        class="accordion__Header-l3iym9-4 cIYXTq"
      >
        Where can I watch?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-l3iym9-2 gwMLOH"
    >
      <div
        class="accordion__Header-l3iym9-4 cIYXTq"
      >
        How do I cancel?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-l3iym9-2 gwMLOH"
    >
      <div
        class="accordion__Header-l3iym9-4 cIYXTq"
      >
        What can I watch on Netflix?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
  </div>
</div>
`;
